ykrt/src/deopt.rs

```diff
@@ -268,18 +277,82 @@ unsafe extern "C" fn __ykrt_deopt(
     guardid: usize,
     // Return address of deoptimize call. Used to find the correct stackmap record.
     retaddr: usize,
+    // The parent trace's call stack at the time of the guard failure. Required to assemble
+    // a side trace.
+    jitcallstack: *const c_void,
     // Current stack pointer. Needed to read spilled register values from the stack.
     rsp: *const c_void,
 ) -> NewFramesInfo {
     // Put the CompiledTrace back into an Arc, so it is dropped properly.
     let ctr = Arc::from_raw(ctr);
 
+    // Check if we have a side trace and execute it.
+    if guardid != SIDETRACE_LAST_GUARD_ID {
+        let guard = &ctr.guards[guardid];
+        if let Some(st) = guard.getct() {
+            let registers = Registers::from_ptr(rsp);
+            let live_vars = ctr.smap.get(&retaddr.try_into().unwrap()).unwrap();
+            let mut ykctrlpvars = Vec::new();
+            for (_i, locs) in live_vars.iter().skip(1).enumerate() {
+                assert!(locs.len() == 1);
```

Comment:
Should probably be `debug_assert`?

Reply:
Given the next line will crash (safely) if this fails, the `assert` isn't doing anything useful. Rather this should be a comment explaining why `locs[0]` won't trigger the `unwrap`.

Reply:
So `let l = locs[0];` will also fail savely, so there's no need for the assert?

Reply:
Correct.

Reply:
Added comment in: 24d4962b8c